### ISSUE TITLE
Deadlock when a task writes to stdout and node reuse is enabled

### ISSUE THREAD

**Issue by rolfbjarne** [dotnet/msbuild#6753]
### Issue Description

MSBuild deadlocks when:

* Executes on macOS (probably linux as well, but probably not Windows)
* A task writes a lot of output to stdout and/or stderr
* This happens in the msbuild server
* 'dotnet build' launched that server

### Steps to Reproduce

Project file:

```xml
<?xml version="1.0" encoding="utf-8"?>
<Project DefaultTargets="Build" ToolsVersion="4.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">

<UsingTask TaskName="Hello"
    TaskFactory="RoslynCodeTaskFactory"
    AssemblyFile="$(MSBuildToolsPath)\Microsoft.Build.Tasks.Core.dll" >
  <ParameterGroup />
  <Task>
    <Code Type="Fragment" Language="cs"><![CDATA[
        var str = "Hello, world!";
        for (int i = 0; i < 100000; i++) {
            Console.WriteLine (str);
        }
        ]]>
    </Code>
  </Task>
</UsingTask>

<Target Name="Build">
  <Hello />
</Target>
</Project>
```

Save the project file (say to `test.proj`), and execute this:

```shell
$ dotnet build-server shutdown 
$ MSBUILDNOINPROCNODE=1 dotnet build test.proj
Microsoft (R) Build Engine version 17.0.0-preview-21403-01+1b4b5fb96 for .NET
Copyright (C) Microsoft Corporation. All rights reserved
[ ... nothing happens ... ]
```

Trying without `MSBUILDNOINPROCNODE=1`:

```shell
$ dotnet build test.proj
Microsoft (R) Build Engine version 17.0.0-preview-21403-01+1b4b5fb96 for .NET
Copyright (C) Microsoft Corporation. All rights reserved.

Hello, world!
Hello, world!
Hello, world!
...
Hello, world!
Hello, world!
Hello, world!

Build succeeded.
    0 Warning(s)
    0 Error(s)

Time Elapsed 00:00:00.95
```

I looked at the code here: https://github.com/dotnet/msbuild/blob/62c6327acdca6af0a2402fd491f91ba5e88d0ede/src/Build/BackEnd/Components/Communications/NodeProviderOutOfProcBase.cs#L495-L501

and it doesn't look like the StandardOutput/StandardError streams are ever read from. This can cause a deadlock, as described in the remarks section here: https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.processstartinfo.redirectstandardoutput?view=net-5.0#remarks

The redirection was implemented here: https://github.com/dotnet/msbuild/commit/dc10ceafe75cdfb01507b63b29a1ecbc7391a8bd, which refers to https://github.com/dotnet/roslyn/pull/23805, which raises this as a concern: https://github.com/dotnet/roslyn/pull/23805/files#r157562697.

The code in `NodeProviderOutOfProcBase.cs` also gave me the idea to try with `MSBUILDENSURESTDOUTFORTASKPROCESSES=1`, and that works just fine:

```shell
$ MSBUILDENSURESTDOUTFORTASKPROCESSES=1 MSBUILDNOINPROCNODE=1 dotnet build test.proj
Microsoft (R) Build Engine version 17.0.0-preview-21403-01+1b4b5fb96 for .NET
Copyright (C) Microsoft Corporation. All rights reserved.

Hello, world!
Hello, world!
Hello, world!
...
Hello, world!
Hello, world!
Hello, world!

Build succeeded.
    0 Warning(s)
    0 Error(s)

Time Elapsed 00:00:01.20
```

```shell
$ dotnet --version
6.0.100-rc.1.21405.1
```

**Comment by rainersigwald** [comment]
Hmm. I wonder if we can just call `BeginOutputReadLine`/`BeginErrorReadLine` and not hook up a delegate to do anything with the data.

If we close stdout in worker-node startup, would that break people who call `Console.WriteLine` in a plugin?

**Comment by BenVillalobos** [comment]
Team Triage: Thanks for filing the issue, how common is this scenario you're running into?

**Comment by rainersigwald** [comment]
Also, is there a particular task that you're seeing that's badly-behaved in this way? They should be logging instead of writing to stdout.

**Comment by filipnavara** [comment]
> Also, is there a particular task that you're seeing that's badly-behaved in this way? They should be logging instead of writing to stdout.

It's an issue with `MergeAppBundles` task in Xamarin. I kind of expected this suggestion. It's possible to wire up the logging but requires a few small refactorings.

Nevertheless the subtlety of the behavior and the resulting deadlock is quite unexpected.

> how common is this scenario you're running into?

The particular task writes a lot of diagnostic output to console. For sufficiently big app I hit the bug in about 95% of the builds.

**Comment by rainersigwald** [comment]
> Nevertheless the subtlety of the behavior and the resulting deadlock is quite unexpected.

Agreed: we should totally fix this! But also ideally use logging in the task :) Note that even when we fix this, the stdout of the worker nodes will go to `/dev/null`, not any console.

**Comment by filipnavara** [comment]
Fixing the particular Xamarin task is doable. I am willing to take a stab at it. However, this was only the tip of the iceberg that tripped the behavior because it had too much output. There's potentially more similar issues so fixing it at MSBuild side would be greatly appreciated.

**Comment by rolfbjarne** [comment]
> Agreed: we should totally fix this!

Great!

> But also ideally use logging in the task :)

Agreed. 

I believe it's also possible to hit this behavior if a task executes another process, doesn't capture stdout/stderr, and that other process writes to stdout/stderr. You could still argue that the task should then capture stdout/stderr from the other process (and redirect to logging), but a plausible explanation could be that the other process doesn't usually write anything, but one day might run into an error condition, and then dumps a lot of information to stderr.

> Note that even when we fix this, the stdout of the worker nodes will go to `/dev/null`, not any console.

That's totally fine.

> Hmm. I wonder if we can just call `BeginOutputReadLine`/`BeginErrorReadLine` and not hook up a delegate to do anything with the data.

I don't see why that wouldn't work, it doesn't matter what you do with the output, you just have to read it.

> If we close stdout in worker-node startup, would that break people who call `Console.WriteLine` in a plugin?

One (admittedly far fetched) failure scenario: someone tries to open stdout, fails unexpectedly (because who considers that opening stdout can fail?), and ends up throwing an exception.

**Comment by Redth** [comment]
@rolfbjarne is this something you can work around in macios tasks for now?  We think this might be the root cause of all of our MAUI CI builds hanging on our own agents, so really blocking for us right now.

**Comment by rolfbjarne** [comment]
@Redth you'd only run into this with our task if you're building an app for more than one iOS runtime identifier (by setting `RuntimeIdentifiers=ios-arm64;ios-arm` for instance). Is that what you're doing?

**Comment by mattleibow** [comment]
I think we are explicitly setting just one.

**Comment by rolfbjarne** [comment]
> I believe it's also possible to hit this behavior if a task executes another process, doesn't capture stdout/stderr, and that other process writes to stdout/stderr.

I believe we ran into this again in a totally different way:

We set the `MONO_ENV_OPTIONS=--trace=E:all` environment variable (to ask mono to print every exception to stdout/stderr), and that ended up hanging msbuild in the same way (sometimes).

**Comment by YuliiaKovalova** [comment]
The issue looks stale. 
Please create a new one if it still remains a problem for you.
